### ISSUE TITLE
Model train

### ISSUE THREAD

**Issue by jaentrouble** [jaentrouble/mouse_chaser_train_2#1]
1. hr538_mouse_food

    - Test run

    - Batchnorm not working properly when training=False

**Comment by jaentrouble** [comment]
2. hr538_mouse_food2

    - Batchnorm gamma = 0.9

    - epoch 10 steps 1000

**Comment by jaentrouble** [comment]
3. hr538_rfcn_mouse_food_1

    - R-FCN implemented

    - epoch 10 steps 1000

**Comment by jaentrouble** [comment]
4. hr538_rfcn_mouse_food2

    - No normalizing layer

    - epoch 100 steps 1000

    - New data added

    - lr: lr_step7_2

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 2000
NMS_TOP_N_RFCN = 100
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.5
NMS_SCORE_THRES_RPN = 0.0
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 0.0
SOFT_SIGMA_RFCN = 1.0

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 1.0

SMOOTH_L1_SIGMA = 1.0
BG_RATIO = 0.75
FG_THRES = 0.5
BBOX_LOSS_GAMMA = 1

ALIGN_RES = 10
OHEM_N = 100

----------------------Epoch 74 abort

**Comment by jaentrouble** [comment]
5. hr538_rfcn_mouse_food3

    - Reload from hr538_rfcn_mouse_food2/7

    - lr: lr_step8_2

    - epoch 100 steps 1000

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 2000
NMS_TOP_N_RFCN = 100
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1 <--------------------- Changed (Does not affect training)
NMS_SCORE_THRES_RPN = 0.4 
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 0.0
SOFT_SIGMA_RFCN = 0.0 <---------------------Changed (Does not affect training)

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 1.0

SMOOTH_L1_SIGMA = 1.0
BG_RATIO = 0.75
FG_THRES = 0.5
BBOX_LOSS_GAMMA = 1
ALIGN_RES = 10
OHEM_N = 100

----------------------------- Epoch 11 abort

**Comment by jaentrouble** [comment]
6. hr538_rfcn_mouse_food4

    - New training

    - lr: lr_step7_2

    - epoch 100 steps 1000

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 2000
NMS_TOP_N_RFCN = 100
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1
NMS_SCORE_THRES_RPN = 0.4 <-------------Changed
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 0.0
SOFT_SIGMA_RFCN = 0.0

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 1.0

SMOOTH_L1_SIGMA = 1.0
BG_RATIO = 0.75
FG_THRES = 0.5
BBOX_LOSS_GAMMA = 1
ALIGN_RES = 10
OHEM_N = 100

-----------------------------Epoch 32 abort

**Comment by jaentrouble** [comment]
7. hr538_rfcn_mouse_food5

    - New training

    - lr: lr_step7_3

    - Normalization: LayerNormalization

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 100 <----------------Changed
NMS_TOP_N_RFCN = 10 <----------------Changed
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1
NMS_SCORE_THRES_RPN = 0.4
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 1.0 <----------------Changed
SOFT_SIGMA_RFCN = 0.0

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 1.0

SMOOTH_L1_SIGMA = 1.0
BG_RATIO = 0.75
FG_THRES = 0.5
BBOX_LOSS_GAMMA = 1

ALIGN_RES = 10 # Crop and resize to (ALIGN_RES*k, ALIGN_RES*k)
OHEM_N = 100

**Comment by jaentrouble** [comment]
### Theories

1. FG_THRES should be low.

    - RoIs are not accurate in the beginning. There won't be enough, if any, foreground rois to learn from, if FG_THRES is high.

    - Another possibility: Train RPN only first, and if it predicts well, fix RPN-backbone and train R-FCN

2. When training, make a lot of RoIs, and pick background RoIs only as much as BG_RATIO

    - Need a lot of RoIs to train r-fcn part. However, just making many RoIs also increase useless background rois too much, leading to bias.

**Comment by jaentrouble** [comment]
8. hr538_rfcn_mouse_food6

    - Load from 7-epoch 100

    - Test Theorem 1.2
    
        - freeze rfcn and backbone part and zero-rfcn loss

        - See if this can improve rpn network

    - lr: low_lr (1e-5)

    - epoch 20 steps 1000

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]
kwargs['frozen_layers'] = ['rfcn', 'backbone']

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 100
NMS_TOP_N_RFCN = 10
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1
NMS_SCORE_THRES_RPN = 0.4
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 1.0
SOFT_SIGMA_RFCN = 0.0

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 0.0 # <-------------- Changed (hr538_rfcn_mf_6)

SMOOTH_L1_SIGMA = 1.0
BG_RATIO = 0.75
FG_THRES = 0.5
BBOX_LOSS_GAMMA = 1

ALIGN_RES = 10 # Crop and resize to (ALIGN_RES*k, ALIGN_RES*k)
OHEM_N = 100

**Comment by jaentrouble** [comment]
### Theories

3. Reg_loss is too small

    - Need to check reg_loss again, but looks like regression is not working well.

    - This may be connected with theorem 1: Only learn from high-RoIs, so regression loss is not big

    - Check reg_loss once more and try increasing reg_loss

### Result (after 9th train)

- Some outliers (ex. [0,1,0,1]) looks like gone, but box doesn't seem to get very close

- For RPN, it only proposes, so regression maybe a bit less important

**Comment by jaentrouble** [comment]
9. hr538_rfcn_mouse_food7

    - load from 8-epoch 20

    - Test Theorem 3

        - Divided BBOX_LOSS_GAMMA to rpn, rfcn each

        - Change BBOX_LOSS_GAMMA_RPN to 10

    - lr: low_lr (1e-5)

    - epoch 50 steps 1000

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]
kwargs['frozen_layers'] = ['rfcn','backbone']

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 100
NMS_TOP_N_RFCN = 10
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1
NMS_SCORE_THRES_RPN = 0.4
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 1.0
SOFT_SIGMA_RFCN = 0.0

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 0.0

SMOOTH_L1_SIGMA = 1.0
BG_RATIO = 0.75
FG_THRES = 0.5
BBOX_LOSS_GAMMA_RPN = 10.0 # <-------------- Changed (hr538_rfcn_mf_7)
BBOX_LOSS_GAMMA_RFCN = 1.0 # <-------------- Changed (hr538_rfcn_mf_7)

ALIGN_RES = 10 # Crop and resize to (ALIGN_RES*k, ALIGN_RES*k)
OHEM_N = 100

**Comment by jaentrouble** [comment]
10. hr538_rfcn_mouse_food8

    - Test Theorem 2

        - Limit background labels' number

    - Unfreeze all layers

    - Load 9-epoch 50

    - lr: low_lr

    - epoch 50 steps 1000

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]
kwargs['frozen_layers'] = []

RPN_TRAIN_THRES = 0.5
BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 100
NMS_TOP_N_RFCN = 10
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1
NMS_SCORE_THRES_RPN = 0.4
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 1.0 
SOFT_SIGMA_RFCN = 0.0

RPN_LOSS_GAMMA = 1.0
RFCN_LOSS_GAMMA = 1.0 # <-----------------Changed(rfcn_mf8)

SMOOTH_L1_SIGMA = 1.0
BG_TRAIN_RATIO = 0.75 # <----------------Name changed
FG_THRES = 0.5
BBOX_LOSS_GAMMA_RPN = 1.0 # <------------Changed(rfcn_mf8)
BBOX_LOSS_GAMMA_RFCN = 1.0 

ALIGN_RES = 10 # Crop and resize to (ALIGN_RES*k, ALIGN_RES*k)
OHEM_N = 100
--------------------------------Abort at epoch 32

**Comment by jaentrouble** [comment]
11. hr538_rfcn_mouse_food9

    - Test Theorem 1.1

        - Freeze all layers and train only rfcn_reg layer

        - Change FG_THRES to 0.2 and SCORE_THRES_RPN to 0.0 -> To make many positive train data

        - Since other layers are all freezed, this won't change classification performance.

        - **EDIT** bbox loss is too low (5e-5), so changed BBOX_LOSS_GAMMA_RFCN to 1.0e3

    - Load 10-epoch 32

    - lr: low_lr

    - epoch 50 steps 1000

kwargs['intermediate_filters'] = 256
kwargs['kernel_size'] = 16
kwargs['stride'] = 8
kwargs['rfcn_window'] = 3
kwargs['anchor_ratios'] = [0.5,1.0,2.0]
kwargs['anchor_scales'] = [0.1,0.3,0.6]
kwargs['frozen_layers'] = ['rpn','rfcn_cls','backbone'] # <---Chaned(rfcn_mf9)

RPN_TRAIN_THRES = 0.5

BATCH_SIZE = 128
POSITIVE_RATIO = 0.5

NMS_TOP_N_RPN = 100
NMS_TOP_N_RFCN = 10
NMS_THRES_RPN = 0.5
NMS_THRES_RFCN = 0.1
NMS_SCORE_THRES_RPN = 0.0 #<---------Changed(rfcn_mf9)
NMS_SCORE_THRES_RFCN = 0.5
SOFT_SIGMA_RPN = 1.0 
SOFT_SIGMA_RFCN = 0.0

RPN_LOSS_GAMMA = 0.0 # <-------------Changed(rfcn_mf9)
RFCN_LOSS_GAMMA = 1.0 

SMOOTH_L1_SIGMA = 1.0

BG_TRAIN_RATIO = 0.75 
FG_THRES = 0.2    #<-----------------Changed(rfcn_mf9)

BBOX_LOSS_GAMMA_RPN = 1.0 
BBOX_LOSS_GAMMA_RFCN = 1.0e3 #<------Changed(rfcn_mf9) 
CLS_LOSS_GAMMA_RPN = 1.0 #<----------New(rfcn_mf9)
CLS_LOSS_GAMMA_RFCN = 0.0 #<---------New(rfcn_mf9)

ALIGN_RES = 10 # Crop and resize to (ALIGN_RES*k, ALIGN_RES*k)
OHEM_N = 100

**Comment by jaentrouble** [comment]
### NOTE

Do not use SOFT_SIGMA : It generally hurts performance, and just costs extra computation.

Do not rely on regression : It does not do well. It is better to make many close rois and choose the best.

- R-FCN has low evaluating cost anyway.

RPN NMS: Its purpose is to decrease cost of r-fcn evaluation.

- Increase RPN_NMS_THRES (iou thres) -> We can rule out with RFCN_NMS; r-fcn has very low cost for evaluating each rois.

RFCN NMS: Its purpose is to erase overlapping boxes

- NMS_SCORE_THRES_RFCN has less meaning. There is 'background' category anyway. 

NMS_TOP_N_RPN = 100
NMS_TOP_N_RFCN = 10
NMS_THRES_RPN = 0.8
NMS_THRES_RFCN = 0.3
NMS_SCORE_THRES_RPN = 0.1 
NMS_SCORE_THRES_RFCN = 0.0
SOFT_SIGMA_RPN = 0.0
SOFT_SIGMA_RFCN = 0.0
